rusoto/credential/src/environment.rs

```diff
@@ -52,24 +43,98 @@ const E_INVALID_EXPIRATION: &str = "Invalid AWS_CREDENTIAL_EXPIRATION in environ
 /// env::set_var("AWS_SESSION_TOKEN", "DfnGs8Td4rT8r4srxAg6Td4rT8r4srxAg6GtkTir");
 ///
-/// let creds = EnvironmentProvider.credentials().wait().unwrap();
+/// let creds = EnvironmentProvider::default().credentials().wait().unwrap();
 ///
 /// assert_eq!(creds.aws_access_key_id(), "ANTN35UAENTS5UIAEATD");
 /// assert_eq!(creds.aws_secret_access_key(), "TtnuieannGt2rGuie2t8Tt7urarg5nauedRndrur");
 /// assert_eq!(creds.token(), &Some("DfnGs8Td4rT8r4srxAg6Td4rT8r4srxAg6GtkTir".to_string()));
 /// assert!(creds.expires_at().is_none()); // doesn't expire
 ///
 /// env::set_var("AWS_CREDENTIAL_EXPIRATION", "2018-04-21T01:13:02Z");
-/// let creds = EnvironmentProvider.credentials().wait().unwrap();
+/// let creds = EnvironmentProvider::default().credentials().wait().unwrap();
 /// assert_eq!(creds.expires_at().unwrap().to_rfc3339(), "2018-04-21T01:13:02+00:00");
 /// # }
 /// ```
-#[derive(Debug)]
-pub struct EnvironmentProvider;
+#[derive(Debug, Clone)]
+pub struct EnvironmentProvider {
+    prefix: String,
+}
+
+impl Default for EnvironmentProvider {
+    fn default() -> Self {
+        EnvironmentProvider {
+            prefix: "AWS".to_owned(),
+        }
+    }
+}
+
+impl EnvironmentProvider {
+    /// Create an EnvironmentProvider with a non-standard variable prefix.
+    ///
+    /// ```rust
+    /// # extern crate futures;
+    /// # extern crate rusoto_credential;
+    /// #
+    /// # fn main() {
+    /// use futures::future::Future;
+    /// use rusoto_credential::{EnvironmentProvider, ProvideAwsCredentials};
+    /// use std::env;
+    ///
+    /// env::set_var("MYAPP_ACCESS_KEY_ID", "ANTN35UAENTS5UIAEATD");
+    /// env::set_var("MYAPP_SECRET_ACCESS_KEY", "TtnuieannGt2rGuie2t8Tt7urarg5nauedRndrur");
+    /// env::set_var("MYAPP_SESSION_TOKEN", "DfnGs8Td4rT8r4srxAg6Td4rT8r4srxAg6GtkTir");
+    ///
+    /// let creds = EnvironmentProvider::with_prefix("MYAPP").credentials().wait().unwrap();
+    ///
+    /// assert_eq!(creds.aws_access_key_id(), "ANTN35UAENTS5UIAEATD");
+    /// assert_eq!(creds.aws_secret_access_key(), "TtnuieannGt2rGuie2t8Tt7urarg5nauedRndrur");
+    /// assert_eq!(creds.token(), &Some("DfnGs8Td4rT8r4srxAg6Td4rT8r4srxAg6GtkTir".to_string()));
+    /// assert!(creds.expires_at().is_none()); // doesn't expire
+    ///
+    /// env::set_var("MYAPP_CREDENTIAL_EXPIRATION", "2018-04-21T01:13:02Z");
+    /// let creds = EnvironmentProvider::with_prefix("MYAPP").credentials().wait().unwrap();
+    /// assert_eq!(creds.expires_at().unwrap().to_rfc3339(), "2018-04-21T01:13:02+00:00");
+    /// # }
+    /// ```
+    pub fn with_prefix(prefix: &str) -> Self {
+        EnvironmentProvider {
+            prefix: prefix.to_owned(),
+        }
+    }
+}
+
+/// A private trait for building the environment variable names based
+/// on a provided prefix. Pulled out of the base `EnvironmentProvider`
+/// to keep separate from the public functionality.
+trait EnvironmentVariableProvider {
```

Comment:
I'm not sure I fully understand the reason you introduced this trait here. This comment says it's to keep the methods separate from the public functionality, and that makes sense to me, but couldn't they just be non-`pub` methods on `EnvironmentProvider` instead?

Reply:
Yeah I went back and forth on whether I wanted to make a Builder object or not. So I compromised and created a minimal trait which would be used to create the variable names. I'm still getting my feet wet in Rust and haven't found a style yet, but I'm kind of a fan of the minimalistic traits and then mix/matching them.

Would modifying the comment be sufficient by explaining that its the smallest possible subset of functionality needed by the `credentials_from_environment` functionality? Or would you like me to migrate the functionality into `EnvironmentProvider` instead?

Reply:
Ah, I understand now, it's just not a pattern I'm familiar with. Personally I'd probably have used a couple of free-standing functions such as `fn access_key_id_var(prefix: &str)`.

I'm okay to leave it as is though, given that it's not a public API I don't think we need to bikeshed it too hard.